### PR TITLE
Support passing parse mode to parse/fingerprint functions

### LINKED ISSUE #215: Parsing queries from plpgsql
Trying to parse queries found in PLpgSQL_expr after calling pg_query_parse_plpgsql. In version 14 and later postgres changed how it returns the query in the struct. Using the examples below, before it would return `SELECT EXISTS(SELECT 1 FROM sample WHERE name = n)` but now it returns the original query `EXISTS(SELECT 1 FROM sample WHERE name = n)` with a parseMode field. Any issues/concerns with exposing parseMode and being able to pass it to parse functions? 

```sql
CREATE OR REPLACE FUNCTION public.function_a(n TEXT) 
RETURNS boolean LANGUAGE plpgsql 
AS 
$$ 
BEGIN 
    IF EXISTS(SELECT 1 FROM sample WHERE name = n) THEN
        RETURN true;
    END IF;
    RETURN false;
END;
$$;
```

#### libpg_query on 13-latest branch
```json
[
    {
        "PLpgSQL_function": {
            "datums": [
                {
                    "PLpgSQL_var": {
                        "refname": "n",
                        "datatype": {
                            "PLpgSQL_type": {
                                "typname": "UNKNOWN"
                            }
                        }
                    }
                },
                {
                    "PLpgSQL_var": {
                        "refname": "found",
                        "datatype": {
                            "PLpgSQL_type": {
                                "typname": "UNKNOWN"
                            }
                        }
                    }
                }
            ],
            "action": {
                "PLpgSQL_stmt_block": {
                    "lineno": 1,
                    "body": [
                        {
                            "PLpgSQL_stmt_if": {
                                "lineno": 1,
                                "cond": {
                                    "PLpgSQL_expr": {
                                        "query": "SELECT EXISTS(SELECT 1 FROM sample WHERE name = n)"
                                    }
                                },
                                "then_body": [
                                    {
                                        "PLpgSQL_stmt_return": {
                                            "lineno": 1,
                                            "expr": {
                                                "PLpgSQL_expr": {
                                                    "query": "SELECT true"
                                                }
                                            }
                                        }
                                    }
                                ]
                            }
                        },
                        {
                            "PLpgSQL_stmt_return": {
                                "lineno": 1,
                                "expr": {
                                    "PLpgSQL_expr": {
                                        "query": "SELECT false"
                                    }
                                }
                            }
                        }
                    ]
                }
            }
        }
    }
]
```

#### libpg_query on 15-latest branch
```json
[
    {
        "PLpgSQL_function": {
            "datums": [
                {
                    "PLpgSQL_var": {
                        "refname": "n",
                        "datatype": {
                            "PLpgSQL_type": {
                                "typname": "UNKNOWN"
                            }
                        }
                    }
                },
                {
                    "PLpgSQL_var": {
                        "refname": "found",
                        "datatype": {
                            "PLpgSQL_type": {
                                "typname": "UNKNOWN"
                            }
                        }
                    }
                }
            ],
            "action": {
                "PLpgSQL_stmt_block": {
                    "lineno": 1,
                    "body": [
                        {
                            "PLpgSQL_stmt_if": {
                                "lineno": 1,
                                "cond": {
                                    "PLpgSQL_expr": {
                                        "query": "EXISTS(SELECT 1 FROM sample WHERE name = n)"
                                    }
                                },
                                "then_body": [
                                    {
                                        "PLpgSQL_stmt_return": {
                                            "lineno": 1,
                                            "expr": {
                                                "PLpgSQL_expr": {
                                                    "query": "true"
                                                }
                                            }
                                        }
                                    }
                                ]
                            }
                        },
                        {
                            "PLpgSQL_stmt_return": {
                                "lineno": 1,
                                "expr": {
                                    "PLpgSQL_expr": {
                                        "query": "false"
                                    }
                                }
                            }
                        }
                    ]
                }
            }
        }
    }
]
```

### PR DESCRIPTION
**Pull request by lcheruka**
Fixes #215

### CODE REVIEW

File: src/pg_query_parse.c [hunk 1]
```diff
@@ -42,7 +42,7 @@ PgQueryInternalParsetreeAndError pg_query_raw_parse(const char* input)
 
 	PG_TRY();
 	{
-		result.tree = raw_parser(input, RAW_PARSE_DEFAULT);
+		result.tree = raw_parser(input, mode);
```
Comment (seanlinsley) [inline src/pg_query_parse.c]: I'm not sure what the fix should be, but there's a warning here:
```
compiling src/pg_query_parse.c
src/pg_query_parse.c:45:35: warning: implicit conversion from enumeration type 'PgQueryParseMode' to different enumeration type 'RawParseMode' [-Wenum-conversion]
                result.tree = raw_parser(input, mode);
                              ~~~~~~~~~~        ^~~~
```

File: test/fingerprint_opts.c [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		}
+		else if (strcmp(result.fingerprint_str, tests[i + 2]) == 0)
+		{
+			printf(".");
+		}
+		else
+		{
+			ret_code = -1;
+			printf("INVALID result for \"%s\" with %d mode\nexpected: \"%s\"\nactual: \"%s\"\nactual tokens: ", tests[i], mode, tests[i + 2], result.fingerprint_str);
+			pg_query_fingerprint_with_opts(tests[i], mode, true);
```
Comment (seanlinsley) [inline test/fingerprint_opts.c:35]: It looks like the `actual tokens: ` should be removed, as well as this `pg_query_fingerprint_with_opts` call. Or did you intend to do something more here?

Comment (lcheruka) [reply test/fingerprint_opts.c:35]: This was copied from the existing fingerprint.c. I believe the second call prints the actual tokens.

**Comment by lfittl** [comment]
Mainly commenting so I don't forget:

For the newly introduced `_opt` functions, I wonder if we should also consider the use cases requested in https://github.com/pganalyze/libpg_query/issues/50 here, i.e. to set GUCs that affect parsing.

Need to think through this more how it could be part of the API, but might be worth doing now so we don't have to break the `_opt` function API in the future.

**Comment by lfittl** [comment]
> Need to think through this more how it could be part of the API, but might be worth doing now so we don't have to break the `_opt` function API in the future.

I've gone ahead and done that change now on this branch (thanks for allowing maintainer edits!), by utilizing an integer as a bitmask instead of a simple enum. From a usability perspective this is essentially the same as if you just care about the parse mode, but if you want to set the options you can just use bitwise OR to enable them.

Overall I think I'm happy with this now. @seanlinsley might be worth if you take another look to see if there are any concerns with that last commit, but otherwise I think we can merge.

**Review by lfittl — APPROVED** [review]

**Review by msepga — APPROVED** [review]

**Review by seanlinsley — APPROVED** [review]

**Comment by lfittl** [comment]
@lcheruka Thanks for the contribution! Merged with slight revisions to allow passing additional options beyond the parse mode.

**Pull request merged**